web/apps/cast/src/services/pair.ts

```diff
@@ -0,0 +1,193 @@
+ */
+export const register = async (): Promise<Registration> => {
+    // Generate keypair.
+    const keypair = await generateKeyPair();
+    const publicKeyB64 = await toB64(keypair.publicKey);
+    const privateKeyB64 = await toB64(keypair.privateKey);
+
+    // Register keypair with museum to get a pairing code.
+    let pairingCode: string;
+    // eslint has fixed this spurious warning, but we're not on the latest
+    // version yet, so add a disable.
+    // https://github.com/eslint/eslint/pull/18286
+    /* eslint-disable no-constant-condition */
+    while (true) {
+        try {
+            pairingCode = await castGateway.registerDevice(publicKeyB64);
+        } catch (e) {
+            log.error("Failed to register public key with server", e);
+        }
+        if (pairingCode) break;
+        // Schedule retry after 10 seconds.
+        await wait(10000);
+    }
+
+    return { pairingCode, publicKeyB64, privateKeyB64 };
+};
+
+/**
+ * Listen for incoming messages on the given {@link cast} receiver, replying to
+ * each of them with a pairing code obtained using the given {@link pairingCode}
+ * callback. Phase 2 of the pairing protocol.
+ *
+ * See: [Note: Pairing protocol].
+ */
+export const advertiseCode = (
+    cast: Cast,
+    pairingCode: () => string | undefined,
+) => {
+    // Prepare the Chromecast "context".
+    const context = cast.framework.CastReceiverContext.getInstance();
+    const namespace = "urn:x-cast:pair-request";
+
+    const options = new cast.framework.CastReceiverOptions();
+    // Do not automatically close the connection when the sender disconnects.
+    options.maxInactivity = 3600; /* 1 hour */
+    // TODO:Is this required? The docs say "(The default type of a message bus
+    // is JSON; if not provided here)."
+    options.customNamespaces = Object.assign({});
+    options.customNamespaces[namespace] =
+        cast.framework.system.MessageType.JSON;
+    // TODO: This looks like the only one needed, but a comment with the reason
+    // might be good.
+    options.disableIdleTimeout = true;
+
+    // Reply with the code that we have if anyone asks over Chromecast.
+    const incomingMessageListener = ({ senderId }: { senderId: string }) => {
+        const code = pairingCode();
+        if (!code) {
+            log.warn(
+                "Ignoring incoming Chromecast message because we do not yet have a pairing code",
+            );
+            return;
+        }
+
+        context.sendCustomMessage(namespace, senderId, { code });
+    };
+
+    context.addCustomMessageListener(
+        namespace,
+        // We need to cast, the `senderId` is present in the message we get but
+        // not present in the TypeScript type.
+        incomingMessageListener as unknown as SystemEventHandler,
+    );
+
+    // Shutdown ourselves if the sender disconnects.
+    // TODO(MR): I assume the page reloads on shutdown. Is that correct?
```

Comment:
@ua741 I've left some TODOs in this file to add a few more comments around various customizations we're doing, if you know the reasons let me know.